### PR TITLE
systemd.network.netdev: add support for WLAN devices

### PR DESCRIPTION
**Pull request by philiptaron**
These are virtual interfaces, where one physical wifi adapter can have more than one interface.

See https://systemd.network/systemd.netdev.html#%5BWLAN%5D%20Section%20Options and https://github.com/systemd/systemd/commit/dedf2d0054eb3150697b1540bf9733cba9e881fd

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/networkd.nix [hunk 1]
```diff
@@ -468,6 +469,19 @@ let
         (assertMinimum "Table" 0)
       ];
 
+      sectionWLAN = checkUnitConfig "WLAN" [
+        (assertOnlyFields [
+          "PhysicalDevice"
+          "Type"
+          "WDS"
+        ])
+        # systemd supports strings here too ("phy0") but index is good enough
```
Comment (arianvp) [inline nixos/modules/system/boot/networkd.nix:478]: Is that confusing for people?

Comment (philiptaron) [reply nixos/modules/system/boot/networkd.nix]: I removed the type check and added a comment.

**Review by ElvishJerricco — APPROVED** [review]
LGTM

**Comment by ElvishJerricco** [comment]
@arianvp do you think we need a nixos test for this? I have no idea if we even *can* test virtual wifi hardware.

**Comment by philiptaron** [comment]
> @arianvp do you think we need a nixos test for this? I have no idea if we even _can_ test virtual wifi hardware.

It looks like it's possible. https://linuxembedded.fr/2021/01/emulating-wlan-in-linux-part-ii-mac80211hwsim

Looks pretty complicated though.

**Pull request merged**